File: apps/api/app/api/dashboards.py
"""Role dashboard routes. TODO I4.

🔴 THE ROLE IS A VIEW, NOT A PRIVILEGE.

`/api/dashboards/{role}` selects WHICH QUESTIONS to answer, not what the
caller is allowed to see. Every query behind it filters `organization_id`
explicitly and runs under RLS and the project-confidentiality predicate, so
asking for the director view does not show a chemist the portfolio — it shows
them the portfolio *they can already reach*, which for somebody on two
projects is two projects.

That is deliberate and it matters. §6 says authorization is on PERMISSIONS and
never on role names, and the frontend's checks are cosmetic. A dashboard that
gated on `role == "director"` would be inventing a second authorization model
beside the real one, in the layer §6 explicitly says must not hold one.

It also makes the product usable: a Lead who wants to see what their chemists
see can open the chemist view, and gets their own data through it.

**The permission floor is `project.view`** — someone with no read access at
all has no business on any of these — and everything past that floor is
decided by the same RLS every other read goes through.
"""

from __future__ import annotations

from typing import Any

from fastapi import APIRouter, Depends, HTTPException, status
from sqlalchemy.orm import Session

from app.agents.orchestrators.root_orchestrator import (
    AgentPrincipal,
    UnknownDashboardError,
    analysis_dashboard,
)
from app.core.security import Principal, get_db, require_permission

router = APIRouter()

__all__ = ["router"]


@router.get("/{role}", tags=["dashboards"])
def get_role_dashboard(
    role: str,
    principal: Principal = Depends(require_permission("project.view")),
    session: Session = Depends(get_db),
) -> dict[str, Any]:
    """One role's dashboard, answered from source records.

    Every panel returns `{available, reason, rows, count}`. A panel whose
    engine does not exist yet comes back `available: false` with the slice
    that will build it — NOT an empty list, because a screen cannot tell
    "nothing to report" from "not built yet" and a reader certainly cannot.

    Every row carries the id of the record it came from, so §2's requirement
    that dashboards "drill down to real source records" is satisfiable by the
    client without a second round trip to work out what it is looking at.
    """
    # 🔴 THROUGH THE ORCHESTRATOR, NOT STRAIGHT INTO THE SERVICE (§0.2, I103).
    #
    # This route imported `app.domains.dashboards.service` and called the
    # builder itself. The analysis conductor existed and NOTHING CALLED IT --
    # a layer with no caller, which is the same defect as a route with no
    # caller, and it meant the "analysis department" was a Python module you
    # could not reach from anywhere in the running product.
    #
    # §0.2 says API routes never call specialists directly. This is the second
    # route to obey it (msd.py was the first), and `tests/test_agent_topology`
    # already forbids importing a conductor here -- so the import is the
    # ORCHESTRATOR, which owns the department gate.
    #
    # ⚠️ `require_permission("project.view")` ABOVE STAYS. The conductor
    # asserts the same permission, and that is deliberate defence in depth,
    # not duplication: this dependency is what refuses an unauthenticated
    # caller before any handler runs, and the conductor is what refuses on the
    # paths that have no route at all. `analysis_conductor.VIEW` is pinned to
    # this literal by a test that reads this file's source, so they cannot
    # drift apart.
    try:
        return analysis_dashboard(
            session,
            name=role,
            # 🔴 ONE ARGUMENT, AND IT CANNOT BE ASSEMBLED FROM VALUES (I104).
            #
            # This used to pass `user_id`, `organization_id` and `permissions`
            # separately, with a comment explaining that the permissions were
            # load-bearing: §11's counts are of ACTIONABLE items, RLS says
            # what may be SEEN and a permission says what may be DONE, and
            # dropping them left the approvals panel offering steps the engine
            # would refuse. All three still reach the builders — they are
            # carried by the principal now, so a caller cannot pass two of the
            # three, and cannot pass a set nobody verified.
            caller=AgentPrincipal.of(principal),
        )
    except UnknownDashboardError as exc:
        # 404 names the four rather than echoing the input back: a message
        # that repeats an arbitrary path segment is a reflected-content
        # smell, and naming the valid set is more useful anyway.
        #
        # Translated HERE rather than left to escape, because the department's
        # error type is not this endpoint's public contract.
        raise HTTPException(
            status_code=status.HTTP_404_NOT_FOUND,
            detail=str(exc),
        ) from exc

File: apps/web/lib/api/dashboards.ts
/**
 * Role dashboards, over HTTP.
 *
 * 🔴 `GET /api/dashboards/{role}` HAD NO BROWSER CALLER AT ALL.
 *
 * The endpoint, four role builders (`chemist`, `engineer`, `lead`,
 * `director`), the analysis conductor behind it and `test_role_dashboards.py`
 * all existed. A `grep` for `api/dashboards` across `apps/web` returned
 * nothing. So every signed-in person saw the same fixed dashboard body no
 * matter who they were — the operator signed in as the **director** and got
 * the chemist's screen — while the sidebar, which is permission-driven, was
 * correctly showing them a director's navigation.
 *
 * That is this project's most-repeated defect: a route with no caller. The
 * live suite could not catch it, because nothing asserted that two roles see
 * different things.
 *
 * ⚠️ THE ROLE IS A VIEW, NOT A PRIVILEGE, AND THAT IS THE SERVER'S RULE.
 *
 * `dashboards.py`'s own header: *"asking for the director view does not show a
 * chemist the portfolio — it shows them the portfolio they can already
 * reach."* Every panel is filtered by RLS and the project-confidentiality
 * predicate regardless of which name is in the path. So choosing a view here
 * is a presentation decision and grants nothing, which is why this module may
 * pick one from the caller's roles without that being an authorization
 * decision in the frontend (§6 forbids those).
 */

import { z } from "zod";

import { apiRequest, type ApiCredentials } from "./client";

/**
 * One panel.
 *
 * 🔴 `available: false` IS NOT AN EMPTY LIST, AND THE SCREEN MUST NOT RENDER
 * IT AS ONE. The server is explicit: a panel whose engine is not built yet
 * comes back unavailable WITH the reason, because "nothing to report" and
 * "not built yet" are different answers and a reader cannot tell them apart
 * from an empty table.
 */
export const dashboardPanelSchema = z.object({
  available: z.boolean(),
  reason: z.string().nullable().optional(),
  // Rows are panel-shaped and differ between panels; each carries the id of
  // the source record so §2's drill-down is satisfiable without a second
  // round trip. Kept loose on purpose rather than enumerated wrongly.
  rows: z.array(z.record(z.string(), z.unknown())).default([]),
  count: z.number().nullable().optional(),
});

export type DashboardPanel = z.infer<typeof dashboardPanelSchema>;

/** The response is an object of named panels; the names differ per role. */
export const roleDashboardSchema = z.record(z.string(), z.unknown());

export type RoleDashboard = Record<string, unknown>;

/**
 * The four the server can build, in the order a person is most likely to want
 * their own view first.
 *
 * 🔴 DERIVED FROM THE SERVER'S `ROLE_DASHBOARDS`, and it must stay in step:
 * asking for a name it does not know is a 404. `dashboards.drift.test.ts`
 * reads the Python registry and fails if these disagree — the same shape as
 * `knowledge.drift.test.ts`, because two literals in two files cannot be
 * type-checked into agreement.
 */
export const DASHBOARD_ROLES = ["chemist", "engineer", "lead", "director"] as const;

export type DashboardRole = (typeof DASHBOARD_ROLES)[number];

/**
 * Which dashboard to open for a caller, from the realm roles they hold.
 *
 * The ten realm roles do not map one-to-one onto the four dashboards, so this
 * is a deliberate, stated mapping rather than a string match that would
 * silently fall through. Most-senior-first: somebody who is both a lead and a
 * chemist gets the lead view, because that is the wider one and they can
 * switch.
 *
 * Returns `null` when the caller holds no role with a dashboard — an
 * executive viewer or a procurement specialist — and the screen then says so
 * rather than showing them a chemist's queue that is not theirs.
 */
export function dashboardForRoles(roles: readonly string[]): DashboardRole | null {
  const held = new Set(roles);
  if (held.has("product_development_director")) return "director";
  if (held.has("product_development_lead")) return "lead";
  if (held.has("product_development_engineer")) return "engineer";
  if (held.has("product_development_chemist")) return "chemist";
  // 🔴 NOT A FALLBACK TO `chemist`. That is precisely the defect being fixed:
  // showing somebody a role's screen that is not theirs. An administrator or
  // an executive viewer has no role dashboard, and saying so is the honest
  // answer.
  return null;
}

export function fetchRoleDashboard(
  credentials: ApiCredentials,
  role: string,
  signal?: AbortSignal,
): Promise<RoleDashboard> {
  return apiRequest(
    { path: `/api/dashboards/${role}`, credentials, signal },
    (payload) => roleDashboardSchema.parse(payload),
  );
}
